### ISSUE TITLE
New Worker not used by previously submitted map

### ISSUE THREAD

**Issue by dotsdl** [dask/distributed#148]
To roughly reproduce, set up a scheduler on one machine with `dscheduler`, start up several workers with `dworker`, and then do

```python
from distributed import Executor

executor = Executor('larry:8786')
L = executor.map(long_running_function, inputs)
```
where the number of `inputs` is greater than the number of processes available among the workers. 

Adding an additional worker with `dworker` later is recognized by the executor, but none of the pending operations appear to be sent to that new worker. New submissions using e.g. `Executor.map`, however, do make use of the new worker.

Not sure if this is technically a bug, but it would be cool if adding more workers helps move pending operations from a `map` further along.

**Comment by mrocklin** [comment]
Agreed.  

The current state is that a task is assigned a worker as soon as all dependencies for that task become available.  It lingers in the stack until it goes to the worker for processing or until the assigned worker dies, at which point it will be appropriately reassigned.

We could consider re-balancing ready tasks on adding a new worker.  We could also consider re-balancing ready tasks whenever a worker becomes idle, which would also help to redistributed straggler tasks.

The only reason we haven't done this yet is that the scheduler doesn't currently hold the data structures necessary to execute this re-balancing operation in linear time relative to the number of moved tasks; so it's potentially costly if we have very many available tasks.  We would need to walk through the tasks on the worker stacks one by one to find good candidates.  It is tricky to add and track additional state to make this more efficient, but doable.

My plan until now has been to handle some other items first.  I'd be happy to re-prioritize though if dynamic scaling is of concrete use.  @dotsdl how meaningfully would this impact your current work?

**Comment by dotsdl** [comment]
I've only been playing with `distributed` for less than a day, so still trying to figure out what it displaces and how in my current workflow. 

Because our lab uses a queueing system (GridEngine) for submitting jobs to our workstations, I'm thinking I'll instantiate workers as individual job submissions whenever I need them, with a scheduler always waiting on my own workstation. The trouble this is that some of those jobs might have to wait a bit while others start pretty soon, in which case the workload starts being directed to only the workers active at the time of the `map` call. Once other workers come online, they won't be utilized unless I issue a different `map` or `submit` call.

So, tl;dr, workers may come on and offline at the whim of the queueing system, and it would be a huge boon if the scheduler handles its workload without fanfare given what workers it has at any moment.

There might be better alternatives to the way I'm planning to use this day-to-day, though. I have to say it's pretty great regardless. :D

**Comment by mrocklin** [comment]
Do the tasks in your workflow have dependencies?  From your perspective what is ideal behavior around stealing tasks assigned to workers because those workers have necessary data.  Should newly arrived workers steal these tasks and transfer data if no tasks-without-dependencies are available for stealing?

**Comment by dotsdl** [comment]
> Do the tasks in your workflow have dependencies?

No, my tasks generally don't have dependencies. My tasks usually store results to disk so they can be retrieved at any time later, so data locality isn't an issue for my particular uses.

Although I can't speak to much depth on the topic, I think that allowing new workers to steal tasks with no dependencies makes definite sense with (relatively) little performance loss (just work for the scheduler). However, no guarantee should be given for this behavior to extend to tasks with dependencies, since moving the data around might be very costly. Not sure what kind of logic would be used to decide what a cutoff is for these dependent-job cases, if any.

Does that sound reasonable/technically feasible?

**Comment by mrocklin** [comment]
Balancing work among tasks without dependencies should be cheap to implement and cheap to compute.  I'm playing with this idea now and I think I have a nice way to implement this that is actually a bit cleaner than the current state.

**Comment by dotsdl** [comment]
Sweet! Happy to give it a try when you've got an experimental branch ready.

**Comment by mrocklin** [comment]
I'm shooting to have something solid by Tuesday.  Other things may come up though.
